File: swift/2.addTwoNumbers.swift
// Using carry flag
func addTwoNumbers(_ l1: ListNode?, _ l2: ListNode?) -> ListNode? {
    let head = ListNode(0)
    var current = head
    var carry = false
    var l1 = l1
    var l2 = l2
    
    while let v1 = l1?.val, let v2 = l2?.val {
        let sum = v1 + v2 + (carry ? 1 : 0)
        carry = sum >= 10
        let new = ListNode(sum % 10)
        current.next = new
        current = new
        
        l1 = l1?.next
        l2 = l2?.next
    }
    while let v1 = l1?.val {
        let sum = v1 + (carry ? 1 : 0)
        carry = sum >= 10
        let new = ListNode(sum % 10)
        current.next = new
        current = new
        
        l1 = l1?.next
    }
    while let v2 = l2?.val {
        let sum = v2 + (carry ? 1 : 0)
        carry = sum >= 10
        let new = ListNode(sum % 10)
        current.next = new
        current = new
        
        l2 = l2?.next
    }
    if carry {
        let new = ListNode(1)
        current.next = new
    }
    
    
    return head.next
}

File: cxx/2.addTwoNumbers.cpp
//
// Created by ALuier Bondar on 2020/10/5.
//

class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode* head = new ListNode(0);
        ListNode* cur = head;
        bool carry = false;
        while (l1 != nullptr && l2 != nullptr) {
            char sum = l1->val + l2->val + (carry ? 1 : 0);
            carry = sum >= 10;
            cur->next = new ListNode(sum % 10);
            cur = cur->next;
            l1 = l1->next;
            l2 = l2->next;
        }
        while (l1 != nullptr) {
            char sum = l1->val + (carry ? 1 : 0);
            carry = sum >= 10;
            cur->next = new ListNode(sum % 10);
            cur = cur->next;
            l1 = l1->next;
        }
        while (l2 != nullptr) {
            char sum = l2->val + (carry ? 1 : 0);
            carry = sum >= 10;
            cur->next = new ListNode(sum % 10);
            cur = cur->next;
            l2 = l2->next;
        }
        if (carry) {
            cur->next = new ListNode(1);
        }
        return head->next;
    }
};